File: web_ui/routes/experience.py
"""Experience tracking routes: summary dashboard and abandoned approach graveyard."""

import json
import math

from starlette.requests import Request
from starlette.responses import HTMLResponse
from starlette.routing import Route

from .. import deps


EFFORTFUL_PAGE_SIZE = 10


async def experience_summary(request: Request) -> HTMLResponse:
    """GET /experience — experience summary dashboard."""
    project = request.query_params.get("project", "")
    outcome_filter = request.query_params.get("outcome", "")
    if outcome_filter not in ("succeeded", "pivoted", "abandoned"):
        outcome_filter = ""
    try:
        page = max(1, int(request.query_params.get("page", "1")))
    except ValueError:
        page = 1

    keys = deps.store.scan_prefix("mem:episodic:")
    all_data = deps.store.get_multi(keys) if keys else []

    total_effort = 0
    count_with_experience = 0
    outcome_counts = {"succeeded": 0, "pivoted": 0, "abandoned": 0}
    effortful = []
    all_abandoned = []
    breakthroughs = []

    for key, data in zip(keys, all_data):
        if data is None:
            continue
        if project and data.get("project") != project:
            continue

        effort_raw = data.get("effort_score")
        if effort_raw is None:
            continue

        try:
            effort = int(float(effort_raw))
        except (ValueError, TypeError):
            continue

        outcome = data.get("outcome", "unknown")
        count_with_experience += 1
        total_effort += effort

        if outcome in outcome_counts:
            outcome_counts[outcome] += 1

        content = (data.get("content") or "")[:100]
        effortful.append({
            "key": key,
            "content": content,
            "effort_score": effort,
            "outcome": outcome,
        })

        # Collect abandoned approaches
        abandoned_raw = data.get("abandoned_approaches", "[]")
        try:
            abandoned = json.loads(abandoned_raw) if abandoned_raw else []
        except (json.JSONDecodeError, TypeError):
            abandoned = []

        for approach in abandoned:
            if isinstance(approach, dict):
                all_abandoned.append({
                    "name": approach.get("name", "?"),
                    "type": approach.get("type", "?"),
                    "reason": approach.get("reason", ""),
                    "effort_score": effort,
                    "memory_key": key,
                })

        # Collect breakthroughs
        breakthrough = data.get("breakthrough")
        if breakthrough:
            breakthroughs.append({
                "key": key,
                "content": content,
                "effort_score": effort,
                "outcome": outcome,
                "breakthrough": breakthrough,
            })

    effortful.sort(key=lambda x: x["effort_score"], reverse=True)
    breakthroughs.sort(key=lambda x: x["effort_score"], reverse=True)

    # Deduplicate graveyard by name
    seen_names = set()
    unique_abandoned = []
    for item in all_abandoned:
        name = item["name"].lower()
        if name not in seen_names:
            seen_names.add(name)
            unique_abandoned.append(item)

    avg_effort = round(total_effort / count_with_experience, 2) if count_with_experience else 0

    # Filter + paginate the effortful table
    if outcome_filter:
        effortful = [m for m in effortful if m["outcome"] == outcome_filter]
    total_pages = max(1, math.ceil(len(effortful) / EFFORTFUL_PAGE_SIZE))
    page = min(page, total_pages)
    start = (page - 1) * EFFORTFUL_PAGE_SIZE
    page_effortful = effortful[start:start + EFFORTFUL_PAGE_SIZE]

    extra_params = ""
    if outcome_filter:
        extra_params += f"&outcome={outcome_filter}"
    if project:
        extra_params += f"&project={project}"

    is_htmx = request.headers.get("HX-Request") == "true"
    template_name = "experience/_effortful.html" if is_htmx else "experience/summary.html"

    template = request.app.state.templates.get_template(template_name)
    content_html = template.render(
        request=request,
        current_page="experience",
        count=count_with_experience,
        avg_effort=avg_effort,
        outcomes=outcome_counts,
        effortful=page_effortful,
        outcome_filter=outcome_filter,
        page=page,
        total_pages=total_pages,
        base_url="/experience",
        extra_params=extra_params,
        breakthroughs=breakthroughs[:5],
        project=project,
    )
    return HTMLResponse(content_html)


async def graveyard(request: Request) -> HTMLResponse:
    """GET /experience/graveyard — abandoned approach graveyard."""
    project = request.query_params.get("project", "")

    keys = deps.store.scan_prefix("mem:episodic:")
    all_data = deps.store.get_multi(keys) if keys else []

    all_abandoned = []
    for key, data in zip(keys, all_data):
        if data is None:
            continue
        if project and data.get("project") != project:
            continue

        effort_raw = data.get("effort_score")
        try:
            effort = int(float(effort_raw)) if effort_raw else None
        except (ValueError, TypeError):
            effort = None

        abandoned_raw = data.get("abandoned_approaches", "[]")
        try:
            abandoned = json.loads(abandoned_raw) if abandoned_raw else []
        except (json.JSONDecodeError, TypeError):
            abandoned = []

        for approach in abandoned:
            if isinstance(approach, dict):
                all_abandoned.append({
                    "name": approach.get("name", "?"),
                    "type": approach.get("type", "?"),
                    "reason": approach.get("reason", ""),
                    "attempted_at": approach.get("attempted_at", ""),
                    "effort_score": effort,
                    "memory_key": key,
                })

    # Deduplicate by name, keeping highest effort
    seen = {}
    for item in all_abandoned:
        name = item["name"].lower()
        if name not in seen or (item["effort_score"] or 0) > (seen[name]["effort_score"] or 0):
            seen[name] = item
    unique_abandoned = sorted(seen.values(), key=lambda x: x.get("effort_score") or 0, reverse=True)

    template = request.app.state.templates.get_template("experience/graveyard.html")
    content_html = template.render(
        request=request,
        current_page="graveyard",
        abandoned=unique_abandoned,
        project=project,
    )
    return HTMLResponse(content_html)


routes = [
    Route("/experience", experience_summary),
    Route("/experience/graveyard", graveyard),
]

File: crates/omnimem-settings/src/experience.rs
//! `/experience` and `/experience/graveyard`: effort, outcomes, breakthroughs
//! and abandoned approaches across episodic memories
//! (`web_ui/routes/experience.py`).

use std::collections::HashMap;
use std::fmt::Write as _;

use axum::extract::{Query, State};
use axum::http::HeaderMap;
use axum::response::Response;
use minijinja::context;
use omnimem_engine::Engine;
use omnimem_engine::pyfmt::round_to;
use omnimem_store::Fields;
use serde_json::{Value, json};

use crate::PanelState;
use crate::pages::{blocking, quote, starting};
use crate::render::page;

const EFFORTFUL_PAGE_SIZE: usize = 10;
const OUTCOMES: [&str; 3] = ["succeeded", "pivoted", "abandoned"];

/// `int(float(raw))`: a whole effort score, when there is a usable one.
fn effort(raw: Option<&String>) -> Option<i64> {
    raw.and_then(|r| r.trim().parse::<f64>().ok())
        .filter(|e| e.is_finite())
        .map(|e| e.trunc() as i64)
}

/// Episodic memories, optionally for one project.
fn episodic(engine: &Engine, project: &str) -> omnimem_engine::Result<Vec<(String, Fields)>> {
    let keys = engine.store().scan_prefix("mem:episodic:")?;
    let rows = engine.store().get_multi(&keys)?;
    Ok(keys
        .into_iter()
        .zip(rows)
        .filter_map(|(key, row)| row.map(|r| (key, r)))
        .filter(|(_, row)| {
            project.is_empty() || row.get("project").map(String::as_str) == Some(project)
        })
        .collect())
}

fn abandoned_approaches(row: &Fields) -> Vec<Value> {
    row.get("abandoned_approaches")
        .filter(|raw| !raw.is_empty())
        .and_then(|raw| serde_json::from_str::<Vec<Value>>(raw).ok())
        .unwrap_or_default()
        .into_iter()
        .filter(Value::is_object)
        .collect()
}

pub(crate) async fn summary(
    State(state): State<PanelState>,
    Query(query): Query<HashMap<String, String>>,
    headers: HeaderMap,
) -> Response {
    let Some(engine) = state.engine() else {
        return starting(&state);
    };
    let project = query.get("project").cloned().unwrap_or_default();
    let outcome_filter = query
        .get("outcome")
        .filter(|o| OUTCOMES.contains(&o.as_str()))
        .cloned()
        .unwrap_or_default();
    let requested_page = query
        .get("page")
        .and_then(|p| p.trim().parse::<i64>().ok())
        .unwrap_or(1)
        .max(1) as usize;

    let scan_project = project.clone();
    let rows = match blocking(move || episodic(&engine, &scan_project)).await {
        Ok(rows) => rows,
        Err(failure) => return failure,
    };

    let mut count = 0usize;
    let mut total_effort = 0i64;
    let mut outcomes: HashMap<&str, usize> = OUTCOMES.iter().map(|o| (*o, 0)).collect();
    let mut effortful: Vec<Value> = Vec::new();
    let mut breakthroughs: Vec<Value> = Vec::new();
    for (key, row) in &rows {
        let Some(effort) = effort(row.get("effort_score")) else {
            continue;
        };
        let outcome = row
            .get("outcome")
            .cloned()
            .unwrap_or_else(|| "unknown".to_owned());
        count += 1;
        total_effort += effort;
        if let Some(n) = outcomes.get_mut(outcome.as_str()) {
            *n += 1;
        }
        let content: String = row
            .get("content")
            .map(|c| c.chars().take(100).collect())
            .unwrap_or_default();
        effortful.push(json!({
            "key": key, "content": content, "effort_score": effort, "outcome": outcome,
        }));
        if let Some(breakthrough) = row.get("breakthrough").filter(|b| !b.is_empty()) {
            breakthroughs.push(json!({
                "key": key, "content": content, "effort_score": effort,
                "outcome": outcome, "breakthrough": breakthrough,
            }));
        }
    }
    let by_effort =
        |a: &Value, b: &Value| b["effort_score"].as_i64().cmp(&a["effort_score"].as_i64());
    effortful.sort_by(by_effort);
    breakthroughs.sort_by(by_effort);
    breakthroughs.truncate(5);

    let avg_effort = if count > 0 {
        json!(round_to(total_effort as f64 / count as f64, 2))
    } else {
        json!(0)
    };
    if !outcome_filter.is_empty() {
        effortful.retain(|m| m["outcome"] == outcome_filter.as_str());
    }
    let total_pages = effortful.len().div_ceil(EFFORTFUL_PAGE_SIZE).max(1);
    let current = requested_page.min(total_pages);
    let effortful: Vec<Value> = effortful
        .into_iter()
        .skip((current - 1) * EFFORTFUL_PAGE_SIZE)
        .take(EFFORTFUL_PAGE_SIZE)
        .collect();

    // Writing into a String cannot fail.
    let mut extra_params = String::new();
    if !outcome_filter.is_empty() {
        let _ = write!(extra_params, "&outcome={}", quote(&outcome_filter));
    }
    if !project.is_empty() {
        let _ = write!(extra_params, "&project={}", quote(&project));
    }
    let template = if headers.get("HX-Request").is_some_and(|v| v == "true") {
        "experience/_effortful.html"
    } else {
        "experience/summary.html"
    };
    page(
        state.templates(),
        template,
        context! {
            current_page => "experience",
            count,
            avg_effort,
            outcomes,
            effortful,
            outcome_filter,
            page => current,
            total_pages,
            base_url => "/experience",
            extra_params,
            breakthroughs,
            project,
        },
    )
}

pub(crate) async fn graveyard(
    State(state): State<PanelState>,
    Query(query): Query<HashMap<String, String>>,
) -> Response {
    let Some(engine) = state.engine() else {
        return starting(&state);
    };
    let project = query.get("project").cloned().unwrap_or_default();
    let scan_project = project.clone();
    let rows = match blocking(move || episodic(&engine, &scan_project)).await {
        Ok(rows) => rows,
        Err(failure) => return failure,
    };

    // One entry per approach name (case-insensitive), keeping the most
    // effortful memory that recorded it.
    let mut unique: Vec<(String, Value)> = Vec::new();
    for (key, row) in &rows {
        let effort = effort(row.get("effort_score").filter(|e| !e.is_empty()));
        for approach in abandoned_approaches(row) {
            let get = |field: &str, default: &str| {
                approach
                    .get(field)
                    .cloned()
                    .unwrap_or_else(|| json!(default))
            };
            let name = get("name", "?");
            let folded = name
                .as_str()
                .map_or_else(|| name.to_string(), str::to_owned)
                .to_lowercase();
            let item = json!({
                "name": name,
                "type": get("type", "?"),
                "reason": get("reason", ""),
                "attempted_at": get("attempted_at", ""),
                "effort_score": effort,
                "memory_key": key,
            });
            match unique.iter_mut().find(|(n, _)| *n == folded) {
                Some((_, kept)) => {
                    if effort.unwrap_or(0) > kept["effort_score"].as_i64().unwrap_or(0) {
                        *kept = item;
                    }
                }
                None => unique.push((folded, item)),
            }
        }
    }
    let mut abandoned: Vec<Value> = unique.into_iter().map(|(_, item)| item).collect();
    abandoned.sort_by(|a, b| {
        b["effort_score"]
            .as_i64()
            .unwrap_or(0)
            .cmp(&a["effort_score"].as_i64().unwrap_or(0))
    });
    page(
        state.templates(),
        "experience/graveyard.html",
        context! { current_page => "graveyard", abandoned, project },
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn effort_scores_truncate_like_int_of_float() {
        assert_eq!(effort(Some(&"4.9".to_owned())), Some(4));
        assert_eq!(effort(Some(&"3".to_owned())), Some(3));
        assert_eq!(effort(Some(&"lots".to_owned())), None);
        assert_eq!(effort(None), None);
    }
}
